Nullable.md

```diff
@@ -0,0 +1,191 @@
+
+Assuming that compiler optimizations are put in place for union types, the
+first step for this Julep is to replace the concrete type `Nullable` with an
+optimized union type. The options for replacing `Nullable{T}` are:
+
+* `Union{T, Void}`
+* `Union{Some{T}, Void}`
+* `Union{T, Null{T}}`
+
+The `Union` type would have the same semantics as unions currently have. Their
+implementation in the scalar case would likely match that of a possible
+`TaggedUnion` type, which represents a discriminated union, but is a concrete
+type that is not a parent of `T`, `Void`, etc.
+
+Deciding between these options requires balancing a variety of concerns. As
+we currently understand them, the arguments are:
+
+### `Union{T, Void}` Pros & Cons
+
+When a function returns `Union{T, Void}`, the user can immediately apply all
+methods defined on `T` without any unwrapping. This is very convenient and
+does not require the introduction of specialized methods.
+
+The counter-argument is that such code will only work if the methods being
+called are defined for both `T` and `Void`. As such, one must redefine
+`f(x::Void) = nothing` for every new function, even though the semantics are
+identical. The requirement for always defining such repetitive functions
+therefore may be perceived as less safe because code that is tested only
+against values of type `T` may fail in untested cases with a method error if
+applied to a value of type `Void`. This could be resolved by static analysis
+tools that check for totality.
+
+### `Union{Some{T}, Void}` Pros & Cons
+
+This approach is essentially the inverse of the `Union{T, Void}`: here no
+methods will work on objects of this type unless they are defined for both
+`f(x::Some{T})` and `f(x::Void)`. As such, this approach requires consistent
+of specialized lifting machinery via dot-broadcasting syntax. See the later
+section on lifting for details.
+
+Because users will have to work with lifted functions, this approach is
+less convenient, but more general and often safer. It also allows one to
+distinguish between outcomes in cases where a method returns something like
+`Union{Some{Void}, Void}` -- as might happen in a hash table implementation
+that returns `Some` when a key is present and associated with `nothing` as a
+value, but returns `Void` when a key is not present.
+
+### `Union{T, Null{T}}` Pros & Cons
+
+This approach allows us to distinguish `Null{Bool}` from `Null{Int}` in
+cases in which that distinction might be useful. R makes such a distinction,
+although it seems to be done only because it is falls out naturally from an
+implementation that uses sentinel values. Otherwise, it is similar to the
+`Union{T, Void}` case described above, except for one crucial problem.
+
+That problem is essentially the problem we already have with `Nullable{T}`:
+writing "clean" functions requires information about type-inference because of
+a coupling of result types to knowledge about counterfactual outcomes in
+branches.
```

Comment:
Is it true that this applies? Wouldn't you use just use dispatch, and define e.g. `&(::Bool, Null{Bool})`?

For containers of `Union{T,Null{T}}`, functions like `map` and `broadcast` will have this reliance on inference just as containers of other types always do currently (it won't matter at all that the container type is `Union{T,Null{T}}`).